### PR TITLE
TST: add test for df.where() with category dtype

### LINKED ISSUE #16979: BUG: DataFrame.where with category dtype
#### Code Sample (it is copy-pastable)

```python
import pandas as pd, numpy as np
df = pd.DataFrame(np.arange(2*3).reshape(2,3), columns=list('abc'))
mask = np.random.rand(*df.shape) < 0.5
df.where(mask)
# Output is correct:
#      a   b    c
# 0  NaN NaN  2.0
# 1  3.0 NaN  NaN

df.a = df.a.astype('category')
df.b = df.b.astype('category')
df.c = df.c.astype('category')
df.where(mask)
# ValueError: Wrong number of items passed 2, placement implies 1
# Expected output: the same as before, but now with dtype `category`.

df.a.where(mask[:,0])
# 0    NaN
# 1    3.0
# Name: a, dtype: float64
# should stay in dtype category

df.a.where(mask[:,0], other=None)
# 0    None
# 1    3
# Name: a, dtype: object
# Expected output: should stay in dtype category
```
#### Problem description

`df.where` should work with all dtypes, the [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.where.html) doesn't say it works only for some dtypes. Also, NaNs are already correctly handled as missing data in `pd.Series` of type 'category', so one should be able to assign NaNs to them. Same with converting the dtype.

While writing this report I found that doing it column-by-column works correctly, so I'll use that as a workaround.

#### Output of ``pd.show_versions()``

<details>
# Paste the output here pd.show_versions() here

INSTALLED VERSIONS                                                                                                                                                                       [1/1839]
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-81-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.2
pytest: None
pip: 9.0.1
setuptools: 36.0.1
Cython: None
numpy: 1.13.1
scipy: 0.19.0
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

Ubuntu  `lsb_release -a`:
------------
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
</details>

### PR DESCRIPTION
**Pull request by ganevgv**
- [x] xref #16979
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_dtypes.py [hunk 1]
```diff
@@ -815,6 +815,19 @@ def test_astype_extension_dtypes_duplicate_col(self, dtype):
...
+        mask = np.array([[True, False, True], [False, True, True]])
+        # change type to category
+        df.A = df.A.astype("category")
+        df.B = df.B.astype("category")
+        df.C = df.C.astype("category")
+
+        expected = df.A.where(mask[:, 0])
+        result = df.A.where(mask[:, 0], other=None)
+        tm.assert_series_equal(result, expected)
+
```
Comment (gfyoung) [inline pandas/tests/frame/test_dtypes.py]: Similar to my comments in https://github.com/pandas-dev/pandas/pull/29455, let's parameterize as follows:

~~~python
@pytest.mark.parametrize("kwargs", [dict(), dict(other=None)])
def test_df_where_with_category(self, kwargs):
    df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
    mask = np.array([[True, False, True], [False, True, True]])

    # Change type to category
    df.A = df.A.astype("category")
    df.B = df.B.astype("category")
    df.C = df.C.astype("category")

    expected = ...   # Construct expected result
    result = df.A.where(mask[:, 0], **kwargs)

    tm.assert_series_equal(result, expected)
~~~

**Comment by gfyoung** [comment]
@ganevgv : Good job with the tests so far!  #16979 is a two-parter, and since both parts have been fixed on `master`, they both need tests.

If you would like, you can add (in this PR) a second test to address the `.astype` error.

**Comment by ganevgv** [comment]
@gfyoung, thanks for the guidance up to now! After adding my second test `test_df_where_change_dtype(...)` in this PR, there seems to be some inconsistencies across the different operating systems.

```
df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
mask = np.array([[True, False, True], [False, True, True]])

result = df.where(mask)
expected = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))

tm.assert_frame_equal(result, expected)
```

After running the code above, on some operating systems (`linux -- Python 3.6.7`, `win32 -- Python 3.6.7`, `win32 -- Python 3.7.5`) applying the boolean mask changes the dtype of the last column `[2, 5]` from `int64` to `int32`. I cannot reproduce this behaviour locally.

`Output of pd.show_versions()`

<details>
INSTALLED VERSIONS
------------------
commit           : 35e91f9fab6f3a9b3d3523d5935b04de888f6daa
python           : 3.6.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 17.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 0.26.0.dev0+734.g0de99558b.dirty
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 39.0.1
Cython           : 0.29.14
pytest           : 5.2.2
hypothesis       : 4.42.6
sphinx           : 2.2.1
blosc            : 1.8.1
feather          : None
xlsxwriter       : 1.2.2
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.9.0
pandas_datareader: None
bs4              : 4.7.1
bottleneck       : 1.2.1
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.0
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
s3fs             : 0.3.5
scipy            : 1.3.1
sqlalchemy       : 1.3.10
tables           : 3.6.1
xarray           : 0.14.0
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.2
</details>

What would be your suggestion? I'm happy to remove or xfail the `test_df_where_change_dtype(...)` test from this PR (so we can merge it) and raise an issue. I can also force the dtype but that's not ideal.

**Comment by gfyoung** [comment]
@ganevgv : Why don't you revert your last commit, and we can focus on it in a separate PR.

**Comment by ganevgv** [comment]
@gfyoung thanks - removed the second test in this PR, will create a new issue specifying the problem

**Comment by gfyoung** [comment]
Actually, can you comment in the original issue?  I plan on keeping that one open.

**Pull request merged**